### ISSUE TITLE
Do project-wide custom assertions really help?

### ISSUE THREAD

**Issue by Reinmar** [ckeditor/ckeditor5#7531]
I'm late to the party. I know that #7513 is closed now, but I needed some time to gather my thoughts.

My current opinion is – project-wide custom assertions are bad due to high cognitive load.

Everyone who has some experience with CKEditor 5 can understand this:

```javascript
expect( selection.getAttribute( 'foo' ) ).to.equal( 'bar' );
```

Everyone can write this too, as you use the basic CKE5 API, that you can also find easily in the API docs.

Now, let's say, we agreed that we use custom assertions for "common checks". You write some test. Is there a custom assertion for what you want to write? Does it fall into "common checks" category? Was it already implemented? What's the syntax? What's the actual behavior of that helper (strict check, loose check, etc)? Where do I actually look for those helpers? Where are they located? Does this helper work with selection or also elements? Does it work with view or model elements?

You know what I'd do? I'd write this:

```javascript
expect( selection.getAttribute( 'foo' ) ).to.equal( 'bar' );
```

Why? Because I don't want to spend 1 minute looking for all the answers. Actually, I tried to find all the answers and I failed – [according to the docs](https://github.com/ckeditor/ckeditor5/pull/7523/files), this helper only handles model selection, but I _think_ it'd work with all the other things too because the API is the same.

It's the same for me with sinon's assertions – I cannot memorize all them. I tend to use chai's `expect()` instead for that reason.

And that's not only about me, of course. I can pretty much guarantee that all contributions that we'll ever get and many our PRs too will not use the custom assertions in some cases. Should we reject them or fix them? That's more work. And, frankly speaking, I often skim tests only looking at test descriptions, not implementation, because that's what matters more to me.

So, I'm sorry for delayed criticism – but to me, this is a wrong direction. Unjustified complication. More code to maintain. More to learn. More to correct in PRs.

---

If you'd like to see this improvement implemented, add a 👍 reaction to this post.

**Comment by Reinmar** [comment]
PS. I specifically called them "project-wide custom assertions". Local helpers do not have the same cognitive load as they are close to the code that uses them and usually do very specific things. Also, if you'll forget to use them, the negative effect on the consistency of the code base is much lower.

**Comment by Reinmar** [comment]
PPS. Of course, these are my thougths, my POV. E.g. I don't take into account (or rather – I don't mention) things like reading results of failed tests where custom assertions could definitely make you faster. That's something I never had that much problem with, so I prioritise this lower than other mentioned issues.

If you see this differently, I won't mind. However, I'd give more focus to documenting well and polishing such assertions (e.g. the model/view or selection/element/text aspects) and documenting them in an easily scannable place (e.g. https://ckeditor.com/docs/ckeditor5/latest/framework/guides/contributing/code-style.html#tests).

**Comment by tomalec** [comment]
> custom assertions are bad due to high cognitive load.

I think it's all about a balance between cognitive load during writing the test vs. cognitive load during debugging of random failure.

Personally, I prefer to minimize the latter. When I write the tests for my change/feature I know the discipline so the decision, like whether it's a property or attribute is relatively easy.

On the other side, when something fails in a distant part of the codebase due to my change, I'd prefer the failure message to deliver me as much information as possible, so it would be easier to me to track the reason of the failure. I'm not writing the code close to that distant part, so the cognitive load on understanding the problem behind "expected false to be true" is relatively bigger.

------


> Everyone who has some experience with CKEditor 5 can understand this:
> ```
> expect( selection.getAttribute( 'foo' ) ).to.equal( 'bar' );
> ```

Sure, I understand this. But the thing is, 

```
 Expected 'baz' to equal "bar"
```
gives me absolutely no clue what failed.
You can as well drop `expect` syntax at all and write
```
const myBoolean = veryComplicatedTestCodeThatsMentToReturnTrue...
assert( myBoolean )
```

-------

>  You write some tests. Is there a custom assertion for what you want to write? Does it fall into "common checks" category? ...

I'd say as it's just a test helper, there is no requirement to use it. You can write is by hand if you want, or use helper if you find one that fits. Just please, pay attention to that someday the test code and failure message will be read by somebody, who is not familiar with this particular test scenario.

-------

> Actually, I tried to find all the answers and I failed – according to the docs, this helper only handles model selection, but I think it'd work with all the other things too because the API is the same.

The assertion works for anything, as `to.equal` works for anything. It simply fails if `anything.getAttribute` fails to respond with given attribute.  

BTW, I'm open for better docs. I believe most of your questions could be easily answered.

-------

> It's the same for me with sinon's assertions – I cannot memorize all them. I tend to use chai's expect() instead for that reason.

If you're not using sinon assertions, in my opinion, there is no point in using `expect` syntax as you do not make use of its benefits over regular `assert`.


----

>  I can pretty much guarantee that all contributions that we'll ever get and many our PRs too will not use the custom assertions in some cases. Should we reject them or fix them? 

In my opinion, if a test does not use any custom assertions but gives meaningful failures, it's totally fine.

-----

> And, frankly speaking, I often skim tests only looking at test descriptions, not implementation, because that's what matters more to me.

That's probably why I have seen some tests that does not test what they describe ;) and failure messages that give more confusion that `undefined is not an object` ;)


------

Speaking of code consistency, lack of custom assertion makes it even worse. I've seen the cases when you have dozens of similar tests for one feature. The tests are added and changed over time. Many of them asserts same thing somewhere inside, but over time the same thing tends to be checked differently, sometimes part's of it become committed, leaving the part untested.

Take, for example, checking that 

```
expect( Array.from( selection.getAttributeKeys() ) ).to.include( "bold" );
expect( selection.isGravityOverridden ).to.true;
//
expect( getSelectionAttributesArray( selection ) ).to.have.members(
	step.selectionAttributes, `#attributes ${ stepString }` );
expect( selection.isGravityOverridden ).to.equal( step.isGravityOverridden, `#isGravityOverridden ${ stepString }` );
// 
expect( [ ...model.document.selection.getAttributeKeys() ] ).to.have.members( [ 'bold' ] );
//
expect( model.document.selection.getAttribute( 'bold' ) ).to.equal( true );
expect( selection.isGravityOverridden ).to.true;
```
Do those differences matter for the sake of the test? was `expect( selection.isGravityOverridden ).to.true;` forgotten or is it not meant to be tested?

----

>  I'd give more focus to documenting well and polishing such assertions (e.g. the model/view or selection/element/text aspects) and documenting them in an easily scannable place (e.g. ckeditor.com/docs/ckeditor5/latest/framework/guides/contributing/code-style.html#tests). Also, rules for what falls under "common case" would need to be defined.

:100: :+1:

-----


Long story short, entire custom assertions ides came to my mind of the frustration of seeing hundreds of meaningless
> Expected true to be false
> Expected spy to be called

from the pieces of code, I have never seen and had full understanding of.

Regardless of the fact I'm relatively new to the project, I think messages like

> Expected model:selection to have attribute "linkHref" of "example.com", but got "github.com"
> Expected formatCallback to be called once with "foo" but it was called thrice

Would speed up my debugging.

**Comment by tomalec** [comment]
Closing in favor of #7538.

The conclusion, from our F2F call, was not to revert this assertion. Keep using it, and in the meantime think about the bigger picture and long term strategy.
